kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k

```diff
@@ -47,36 +48,28 @@ module BITWISE-SIMPLIFICATION [symbolic]
-    rule notMaxUInt32  &Int X => 0 requires #rangeUInt( 32, X) [simplification]
-    rule notMaxUInt64  &Int X => 0 requires #rangeUInt( 64, X) [simplification]
-    rule notMaxUInt96  &Int X => 0 requires #rangeUInt( 96, X) [simplification]
-    rule notMaxUInt128 &Int X => 0 requires #rangeUInt(128, X) [simplification]
-    rule notMaxUInt160 &Int X => 0 requires #rangeUInt(160, X) [simplification]
-    rule notMaxUInt192 &Int X => 0 requires #rangeUInt(192, X) [simplification]
-    rule notMaxUInt208 &Int X => 0 requires #rangeUInt(208, X) [simplification]
-    rule notMaxUInt224 &Int X => 0 requires #rangeUInt(224, X) [simplification]
-    rule notMaxUInt240 &Int X => 0 requires #rangeUInt(240, X) [simplification]
-    rule notMaxUInt248 &Int X => 0 requires #rangeUInt(248, X) [simplification]
-
-    rule notMaxUInt240 &Int (X <<Int 240) => X <<Int 240 requires #rangeUInt(16, X) [simplification]
-    rule notMaxUInt248 &Int (X <<Int 248) => X <<Int 248 requires #rangeUInt( 8, X) [simplification]
-
-    rule maxUInt8   &Int (X |Int (notMaxUInt8   &Int _)) => maxUInt8   &Int X [simplification]
-    rule maxUInt16  &Int (X |Int (notMaxUInt16  &Int _)) => maxUInt16  &Int X [simplification]
-    rule maxUInt32  &Int (X |Int (notMaxUInt32  &Int _)) => maxUInt32  &Int X [simplification]
-    rule maxUInt64  &Int (X |Int (notMaxUInt64  &Int _)) => maxUInt64  &Int X [simplification]
-    rule maxUInt96  &Int (X |Int (notMaxUInt96  &Int _)) => maxUInt96  &Int X [simplification]
-    rule maxUInt128 &Int (X |Int (notMaxUInt128 &Int _)) => maxUInt128 &Int X [simplification]
-    rule maxUInt160 &Int (X |Int (notMaxUInt160 &Int _)) => maxUInt160 &Int X [simplification]
-    rule maxUInt192 &Int (X |Int (notMaxUInt192 &Int _)) => maxUInt192 &Int X [simplification]
-    rule maxUInt208 &Int (X |Int (notMaxUInt208 &Int _)) => maxUInt208 &Int X [simplification]
-    rule maxUInt224 &Int (X |Int (notMaxUInt224 &Int _)) => maxUInt224 &Int X [simplification]
-    rule maxUInt240 &Int (X |Int (notMaxUInt240 &Int _)) => maxUInt240 &Int X [simplification]
-    rule maxUInt248 &Int (X |Int (notMaxUInt248 &Int _)) => maxUInt248 &Int X [simplification]
+    // Generalization of: notMaxUIntXXX &Int Y => 0 requires #rangeUInt(XXX, Y)
+    rule [bitwise-and-notMaxUInt-zero]:
+      X &Int Y => 0
+      requires #rangeUInt(256, X)
+       andBool 2 ^Int ( log2Int ( pow256 -Int X ) ) ==Int pow256 -Int X
+       andBool 0 <=Int Y andBool Y <Int pow256 -Int X
+      [simplification, concrete(X)]
+
+    // Generalization of: notMaxUIntXXX &Int (Y <<Int XXX) => Y <<Int XXX requires #rangeUInt(256 -Int XXX, Y)
+    rule X &Int (Y <<Int Z) => Y <<Int Z
+      requires #rangeUInt(256, X)
+       andBool 2 ^Int ( log2Int ( pow256 -Int X ) ) ==Int pow256 -Int X
+       andBool Z ==Int log2Int ( pow256 -Int X )
+       andBool 0 <=Int Y andBool Y <Int 2 ^Int ( 256 -Int Z )
+      [simplification, concrete(X, Z), comm]
```

Comment:
Can this be generalized further to be something like:

```
rule X &Int Y => Y
  requires #rangeUInt(256, X) andBool #rangeUInt(256, Y)
  andBool 2 ^Int (log2Int(pow256 -Int X)) ==Int pow256 -Int X
  andBool Y -Int (pow256 -Int X) ==Int Y
```

The last condition ensures that `Y` has no bits set lower than `X` does.

Then we probably need another lemma about `<<Int` which says:

```
rule (Y <<Int Z) -Int (pow256 -Int X) => Y <<Int Z
  requires #rangeUInt(256, X) andBool #rangeUInt(256, Y) andBool SOME OTHER CONDITIONS
```

Not sure this is a better approach, just having ideas.

Reply:
Another question though, is `Z ==Int log2Int(pow256 -Int X)` required, or is `log2Int(pow256 -Int X) <=Int Z` good enough?

Reply:
I'm not quite understanding. Doesn't it hold that:
```
Y -Int (pow256 -Int X) ==Int Y  <==> [cancel out Y]
0 -Int (pow256 -Int X) ==Int 0  <==> [-Int distributivity]
X -Int pow256 ==Int 0           <==> [pow256 to the RHG]
X ==Int pow256
```
I do see the idea, though, I will think about it, it does make sense.

Reply:
I think `log2Int(pow256 -Int X) <=Int Z ` is good enough, yes, but then I would have also ask for `Z <Int 256`, I think.